### ISSUE TITLE
The deadlock test case in 'main.m' doesn't work.

### ISSUE THREAD

**Issue by alexlee002** [ccgus/fmdb#598]
I don't know how to simulate a deadlock case

**Comment by ccgus** [comment]
Interesting, I wonder if something changed in the sqlite library to keep this from happening. It _used_ to work.

**Comment by alexlee002** [comment]
@ccgus  YES, it did work years ago.   I met a big performance issue while using FMDB queue,  so I plan to use FMDB pool, so I need to avoid deadlock case, and a better way to handle the SQLITE_BUSY.

**Comment by groue** [comment]
You may both be interested in GRDB's DatabasePool. It's in Swift, but nothing prevents it from being ported to Objective-C.

DatabasePool is based on SQLite [WAL mode](https://sqlite.org/wal.html). It allows safe concurrent reads and writes. Writes are serialized. Writes don't block reads. Several reads can be performed in parallel. Reads are guaranteed a consistent and immutable view of the last committed state of the database. There are no SQLITE_BUSY errors, ever.

To achieve this, a database pool maintains a single serialized writer connection, and a pool of several serialized read-only connections. The use of several serialized read-only connection instead of a single concurrent connection allows readers to be isolated from concurrent writes by the mean of deferred transactions. Those automatic deferred transactions are the cornerstone of an understandable API which does not rely on the user to write their own safety belt around SQLite concurrency (since this is very hard). And since "safety" should not mean "toy library", DatabasePool allows *precise* scheduling between writes and reads so that you never block a dispatch queue longer than necessary.

I'd be happy discussing the benefits of DatabasePool with @ccgus and @robertmryan if you are ever interested adding support for WAL mode in FMDB 3.

- implementation: https://github.com/groue/GRDB.swift/blob/v1.1/GRDB/Core/DatabasePool.swift
- documentation: https://github.com/groue/GRDB.swift/blob/v1.1/README.md#database-pools
- concurrency in GRDB: https://github.com/groue/GRDB.swift/blob/v1.1/README.md#concurrency
